### PR TITLE
Add some details to cherry picking for newcomers

### PR DESCRIPTION
**Pull request by jayunit100**
@pweil- this would probably save some folks some time.  just jots down some notes from today in irc #11582

### CODE REVIEW

File: HACKING.md [hunk 1]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
 
 If you are not updating packages you should not need godep installed.
 
-## Cherry-picking an upstream commit into Origin
+## Cherry-picking an upstream commit into Origin: Why, how, and when.
+
+Upstream origin carries patches inside of vendor/ on top of each rebase.  Thus, Origin consumes kubernetes in two ways.
+
+1. Periodic rebases against a kubernetes commit.  Eventually, any patch you have in upstream kubernetes will land in openshift via this mechanism.
```
Comment (ncdc) [inline HACKING.md]: s/patch/code/

File: HACKING.md [hunk 2]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+2. Cherry picked patches for important features.  This is done when an important feature/fix is in a version of kubernetes that we have not yet rebased against.
```
Comment (ncdc) [inline HACKING.md]: "Cherry-picked patches for important bug fixes". We're trying to limit feature backporting to 0 if possible.

File: HACKING.md [hunk 3]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+### Manually
+
+You can manually try to cherry pick a commit, by using git apply. This can easily be done in a couple of steps.
```
Comment (ncdc) [inline HACKING.md]: comma unnecessary

File: HACKING.md [hunk 4]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+- wget the patch, i.e. `wget -O /tmp/mypatch https://github.com/kubernetes/kubernetes/pull/34624.patch`
+- PATCH=/tmp/mypatch git apply --directory vendor/k8s.io/kubernetes $PATCH
+
+If this fails, then its possible you may need to pick multiple commits.
+
+### For openshift newcomers: Should I cherry pick my kubernetes fix into openshift? Or should I wait for the next rebase ?
+
+If youre patch is really far behind, for example, if there have been 5 commits modifying the directory you care about, cherry picking will be increasingly difficult and you should consider waiting for the next rebase, which will likely include the commit you care about, or at least decrease the amount of cherry picks you need to do to merge.
```
Comment (ncdc) [inline HACKING.md]: s/youre/your/

File: HACKING.md [hunk 5]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+To really know the answer, you need to know *how many commits behind you are in a particular directory*, often.  To do this, just use git log, like so (using pkg/scheduler/ as an example).
+
+```
+MYDIR=pkg/scheduler/algorithm git log --oneline -- vendor/k8s.io/kubernetes/${MYDIR} | grep UPSTREAM | cut -d' ' -f 4-10 | head -1
+```
+
+The commit message printed above will tell you what the LAST commit in kubernetes was which happened in the "/pkg/scheduler/algorithm" directory, which will give you an intuition about how "hot" the code you are cherry picking is.  If it has changed alot, recently, then that means you probably will want to wait for a rebase to land.
```
Comment (ncdc) [inline HACKING.md]: s/alot/a lot/

File: HACKING.md [hunk 6]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+### Using hack/cherry-pick
+
+For convenience, you can use `hack/cherry-pick.sh` to generate patches for Origin from upstream commits.
+
+The purpose of this command is to allow you to pull individual commits from a local kubernetes repository into origins vendored kuberenetes in a fully automated manner.
```
Comment (ncdc) [inline HACKING.md]: s/origins/origin's/
typo: kuberenetes

File: HACKING.md (hunk 1 above)
Comment (pweil-) [inline HACKING.md]: Please caps for Kubernetes and OpenShift

File: HACKING.md [hunk 7]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+1. Periodic rebases against a kubernetes commit.  Eventually, any patch you have in upstream kubernetes will land in openshift via this mechanism.
+
+2. Cherry picked patches for important features.  This is done when an important feature/fix is in a version of kubernetes that we have not yet rebased against.
+
+### Manually
+
+You can manually try to cherry pick a commit, by using git apply. This can easily be done in a couple of steps.
+
+- wget the patch, i.e. `wget -O /tmp/mypatch https://github.com/kubernetes/kubernetes/pull/34624.patch`
+- PATCH=/tmp/mypatch git apply --directory vendor/k8s.io/kubernetes $PATCH
```
Comment (pweil-) [inline HACKING.md]: Not sure about others but I always let the patch go to <pr number>.patch in my current directory and apply there.  Using the output file and then a subsequent env var seems workable but maybe not what I'd consider normal.  I don't feel strongly - this definitely works, just caught my eye.

File: HACKING.md [hunk 8]
```diff
@@ -224,10 +224,51 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+If this fails, then its possible you may need to pick multiple commits.
+
+### For openshift newcomers: Should I cherry pick my kubernetes fix into openshift? Or should I wait for the next rebase ?
```
Comment (pweil-) [inline HACKING.md]: extra space at the end here

File: HACKING.md [hunk 9]
```diff
@@ -224,10 +224,57 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+2. Cherry-picked patches for important *bug fixes*.  We really try to limit feature back-porting entirely.
+
+### Manually
+
+You can manually try to cherry pick a commit (by using git apply). This can easily be done in a couple of steps.
+
+- wget the patch, i.e. `wget -O /tmp/mypatch https://github.com/kubernetes/kubernetes/pull/34624.patch`
+- PATCH=/tmp/mypatch git apply --directory vendor/k8s.io/kubernetes $PATCH
+
+If this fails, then its possible you may need to pick multiple commits.
```
Comment (pweil-) [inline HACKING.md]: s/its/it's

File: HACKING.md [hunk 10]
```diff
@@ -224,10 +224,57 @@ OpenShift is checked into this repository.  To install `godep` locally run:
...
+
+### For Openshift newcomers: Should I cherry pick my kubernetes fix into Openshift ? Or should I wait for the next rebase?
+
+Assuming you read the bullets above... If your patch is really far behind, for example, if there have been 5 commits modifying the directory you care 
+
```
Comment (pweil-) [inline HACKING.md]: Looks like this section got some extra line breaks that are messing up the formatting if you look at the rendered markdown.

File: HACKING.md [hunk 11]
```diff
@@ -224,10 +224,57 @@ OpenShift is checked into this repository.  To install `godep` locally run:
 
 If you are not updating packages you should not need godep installed.
 
-## Cherry-picking an upstream commit into Origin
+## Cherry-picking an upstream commit into Origin: Why, how, and when.
+
+Upstream origin carries patches inside of vendor/ on top of each rebase.  Thus, Origin consumes Kubernetes in two ways.
```
Comment (pweil-) [inline HACKING.md]: s/origin/Origin

I think you mean "Origin carries upstream patches" rather than "Upstream origin carries patches"

**Comment by jayunit100** [comment]
ok done i think,

**Comment by jayunit100** [comment]
@pweil- all set on this ?

**Comment by pweil-** [comment]
yes, all set from my feedback

**Comment by jayunit100** [comment]
ok ill see if i can clean the lines on this doc to

**Comment by jayunit100** [comment]
@ncdc fixed the formatting for the whole doc in a separate commit.  LG2U ?

**Comment by ncdc** [comment]
I probably would have preferred a full reformat in a separate PR, as it makes it hard to see what you've changed specific to this one.

**Comment by pweil-** [comment]
[merge]

**Pull request merged**